### ISSUE TITLE
const enum keyof typeof 

### ISSUE THREAD

**Issue by styfle** [Microsoft/TypeScript#21031]
**TypeScript Version:**  2.7.0-dev.20180105

**Code** `index.ts`

```ts
const enum FileType { png, jpg }
type FileTypeStr = keyof typeof FileType; // fails, not expected!

function f1(f: FileType) { }
function f2(f: FileTypeStr) { }

f1(FileType.png); // works as expected
f1(FileType.nope); // fails as expected
f2('png'); // works as expected
f2('nope'); // fails as expected
```

**Expected behavior:**

Expected the second line to work without compilation errors just like a non-const enum.

**Actual behavior:**

Emits an unexpected error on line 2:

```
index.ts(2,33): error TS2475: 'const' enums can only be used in property or index access expressions or the right hand side of an import declaration or export assignment.
```

Oddly enough, it looks like the type checking below the type declaration is working as expected. So I think this might be bug...line 2 should not be a compiler error.

Related issue: #14106

**Comment by sylvanaar** [comment]
Seems like it should be allowed in a TypeQuery.

**Comment by mhegazy** [comment]
PRs welcomed.

**Comment by styfle** [comment]
@mhegazy I can try taking this on but I don't know where to start.
I found error code 2475 [diagnosticMessages.json](https://github.com/Microsoft/TypeScript/blob/73e3e8d790410c982ea80fd786cb046b0b861989/src/compiler/diagnosticMessages.json) but where do I go from here?

**Comment by Kingwl** [comment]
@styfle   
I'm working on this  
(and in fact it's almost done

**Comment by mhegazy** [comment]
thanks @Kingwl !

**Comment by styfle** [comment]
Great job @Kingwl and @mhegazy! 🎉 

When can I expect this to be released?
Which milestone does this fall under?

**Comment by mhegazy** [comment]
https://github.com/Microsoft/TypeScript/wiki/Roadmap#27-january-2018
